cmd/vsphere-cloud-controller-manager/main.go

```diff
@@ -191,11 +193,23 @@ func main() {
 			_ = watch.Close() // ignore explicitly when the watch closes
 		}(watch)
 
+		// Notify the stop channel on SIGTERM or SIGINT in order
+		// to run cleanup such as logout of VSphere sessions
+		cancelChan := make(chan os.Signal, 1)
+		signal.Notify(cancelChan, syscall.SIGTERM, syscall.SIGINT)
+		go func() {
+			sig := <-cancelChan
+			klog.Infof("Signal received: %s. Stopping...\n", sig)
+			close(stop)
+		}()
+
 		if err := app.Run(completedConfig, cloud, controllerInitializers, webhookHandlers, stop); err != nil {
 			// explicitly ignore the error by Fprintf, exiting anyway due to app error
 			_, _ = fmt.Fprintf(os.Stderr, "%v\n", err)
 			os.Exit(1)
 		}
+		// Log out of all sessions on exit or signal
+		vsphere.SessionLogout()
```

Comment:
Question: This one is added here, not in the goroutine, is it because it needs to notify the init() func to logout all sessions and can block the main.go? Just curious, if app.Run exit with error will this code still get executed?

Reply:
It turns out if there are errors after app initialization, the program just exits without giving us a chance to cleanup sessions. This is due to the Fatalf handling of errors k8s.io/cloud-provider, so it's a bit difficult to fix that at the moment.